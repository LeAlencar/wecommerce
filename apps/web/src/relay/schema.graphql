"""
Exposes a URL that specifies the behavior of this scalar.
"""
directive @specifiedBy(
  """
  The URL that specifies the behavior of this scalar.
  """
  url: String!
) on SCALAR

"""
Charge type
"""
type Charge implements Node {
  """
  The ID of an object
  """
  id: ID!

  """
  Charge brCode
  """
  brCode: String!

  """
  Charge's customerName
  """
  customerName: String!

  """
  Charge's customerTaxID
  """
  customerTaxID: String!

  """
  Charge's customerEmail
  """
  customerEmail: String!

  """
  Charge's productID
  """
  product: Product!
}

"""
A connection to a list of items.
"""
type ChargeConnection {
  """
  Information to aid in pagination.
  """
  pageInfo: PageInfo!

  """
  A list of edges.
  """
  edges: [ChargeEdge]
}

input ChargeCreateInput {
  customerName: String!
  customerTaxID: String!
  customerEmail: String!
  product: String!
  clientMutationId: String
}

type ChargeCreatePayload {
  error: String
  node: Charge
  success: String
  clientMutationId: String
}

"""
An edge in a connection.
"""
type ChargeEdge {
  """
  The item at the end of the edge
  """
  node: Charge

  """
  A cursor for use in pagination
  """
  cursor: String!
}

"""
Root of ... mutations
"""
type Mutation {
  userCreate(input: UserRegisterInput!): UserRegisterPayload
  userLogin(input: UserLoginInput!): UserLoginPayload
  userLogout(input: UserLogoutInput!): UserLogoutPayload

  """
  add a new product
  """
  ProductAddMutation(input: ProductAddInput!): ProductAddPayload

  """
  dele a product
  """
  ProductDeleteMutation(input: ProductDeleteInput!): ProductDeletePayload

  """
  create a new charge
  """
  ChargeCreateMutation(input: ChargeCreateInput!): ChargeCreatePayload
}

"""
An object with an ID
"""
interface Node {
  """
  The id of the object.
  """
  id: ID!
}

"""
Information about pagination in a connection.
"""
type PageInfo {
  """
  When paginating forwards, are there more items?
  """
  hasNextPage: Boolean!

  """
  When paginating backwards, are there more items?
  """
  hasPreviousPage: Boolean!

  """
  When paginating backwards, the cursor to continue.
  """
  startCursor: String

  """
  When paginating forwards, the cursor to continue.
  """
  endCursor: String
}

"""
Product type
"""
type Product implements Node {
  """
  The ID of an object
  """
  id: ID!

  """
  Product's name
  """
  name: String!

  """
  Product's displayName
  """
  displayName: String!

  """
  Product's description
  """
  description: String!

  """
  Product's price
  """
  price: Int!
}

input ProductAddInput {
  name: String!
  displayName: String!
  description: String!
  price: Int!
  clientMutationId: String
}

type ProductAddPayload {
  productEdge: ProductEdge
  error: String
  clientMutationId: String
}

"""
A connection to a list of items.
"""
type ProductConnection {
  """
  Information to aid in pagination.
  """
  pageInfo: PageInfo!

  """
  A list of edges.
  """
  edges: [ProductEdge]
}

input ProductDeleteInput {
  id: String!
  clientMutationId: String
}

type ProductDeletePayload {
  productId: ID
  error: String
  success: String
  clientMutationId: String
}

"""
An edge in a connection.
"""
type ProductEdge {
  """
  The item at the end of the edge
  """
  node: Product

  """
  A cursor for use in pagination
  """
  cursor: String!
}

"""
The root of all queries
"""
type Query {
  """
  Fetches an object given its ID
  """
  node(
    """
    The ID of an object
    """
    id: ID!
  ): Node

  """
  Fetches objects given their IDs
  """
  nodes(
    """
    The IDs of objects
    """
    ids: [ID!]!
  ): [Node]!
  products(
    after: String
    first: Int
    before: String
    last: Int
    id: String
  ): ProductConnection!
  charges(
    after: String
    first: Int
    before: String
    last: Int
    id: String
  ): ChargeConnection!
  me: User
  version: String
}

"""
User type
"""
type User implements Node {
  """
  The ID of an object
  """
  id: ID!

  """
  User's username
  """
  username: String!

  """
  User's email
  """
  email: String!
}

input UserLoginInput {
  email: String!
  password: String!
  clientMutationId: String
}

type UserLoginPayload {
  token: String
  me: User
  clientMutationId: String
}

input UserLogoutInput {
  clientMutationId: String
}

type UserLogoutPayload {
  error: String
  success: String
  clientMutationId: String
}

input UserRegisterInput {
  username: String!
  pixKey: String!
  email: String!
  password: String!
  clientMutationId: String
}

type UserRegisterPayload {
  user: String
  token: String
  error: String
  clientMutationId: String
}
