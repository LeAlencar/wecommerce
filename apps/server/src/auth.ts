import jwt from 'jsonwebtoken';
import { jwtSecret } from './config';
import UserModel, { IUser } from './modules/user/UserModel';

export const getUser = async (token: string | null | undefined) => {
  if (!token) return { user: null };

  try {
    const decodedToken = jwt.verify(
      token?.replace('JWT', '').trim(),
      jwtSecret
    );

    const user = await UserModel.findOne({
      _id: (decodedToken as { id: string }).id
    });

    return {
      user
    };
  } catch (err) {
    console.log(err);
    return { user: null };
  }
};

export const generateJwtToken = (user: IUser) => {
  return `JWT ${jwt.sign({ id: user._id }, `jwtsupersecreto`)}`;
};
