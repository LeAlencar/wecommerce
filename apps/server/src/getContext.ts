import { IUser } from './modules/user/UserModel';
import { Context } from 'koa';
import { getDataloaders } from './modules/loader/loaderRegister';
import { GraphQLContext } from './types/types';
import { Request as HelixRequest } from 'graphql-helix';
import { load } from './modules/user/UserLoader';

type ContextVars = {
  user?: IUser | null;
  req?: HelixRequest;
  koaContext: Context;
  setCookie: (cookieName: string, token: string) => void;
};

export const getContext = async (ctx: ContextVars) => {
  const context = {
    ...ctx
  };
  const dataloaders = getDataloaders();

  if (context.user) {
    context.user = await load({ ...context, dataloaders }, context.user._id);
  }

  return {
    req: ctx.req,
    dataloaders,
    user: ctx.user,
    koaContext: ctx.koaContext,
    setCookie: ctx.setCookie
  } as GraphQLContext;
};
