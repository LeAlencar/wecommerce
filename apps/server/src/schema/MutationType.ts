import { GraphQLObjectType } from 'graphql';

import UserMutations from '../modules/user/mutations';
import ProductMutations from '../modules/product/mutations';
import ChargeMutations from '../modules/charge/mutations';

export default new GraphQLObjectType({
  name: 'Mutation',
  description: 'Root of ... mutations',
  fields: () => ({
    ...UserMutations,
    ...ProductMutations,
    ...ChargeMutations
  })
});
