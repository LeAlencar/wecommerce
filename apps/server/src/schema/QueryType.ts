import { GraphQLObjectType, GraphQLString } from 'graphql';
import { UserType } from '../modules/user/UserType';
import { nodeField, nodesField } from '../modules/node/typeRegister';
import * as UserLoader from '../modules/user/UserLoader';
import { productConnectionField } from '../modules/product/ProductConnectionField';
import pkg from '../../package.json';

const QueryType = new GraphQLObjectType({
  name: 'Query',
  description: 'The root of all queries',
  fields: () => ({
    node: nodeField,
    nodes: nodesField,
    ...productConnectionField(),
    me: {
      type: UserType,
      resolve: async (_, args, context) => {
        return await UserLoader.load(context, context.user?._id);
      }
    },
    version: {
      type: GraphQLString,
      resolve: () => pkg.version
    }
  })
});
export default QueryType;
