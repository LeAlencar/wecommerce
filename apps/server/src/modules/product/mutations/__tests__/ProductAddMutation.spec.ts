import {
  clearDbAndRestartCounters,
  connectMongoose,
  defaultFrozenKeys,
  disconnectMongoose,
  sanitizeTestObject
} from '@wecommerce/test';
import { generateJwtToken } from '../../../../auth';
import { getContext } from '../../../../getContext';
import { graphql } from 'graphql';
import { schema } from '../../../../schema/schema';
import { createUser } from '../../../user/fixture/createUser';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it('should create an product', async () => {
  const user = await createUser();
  generateJwtToken(user);

  const query = /* GraphQL */ `
    mutation productAddMutation($input: ProductAddInput!) {
      ProductAddMutation(input: $input) {
        productEdge {
          node {
            id
            name
          }
          cursor
        }
        error
      }
    }
  `;

  const input = {
    name: 'Awesome-Product',
    displayName: 'Awesome Product',
    description: 'awesome product',
    price: 1000
  };

  const variables = {
    input
  };

  const context = await getContext({
    user
  });

  const rootValue = {};

  const result = await graphql({
    schema: schema,
    source: query,
    rootValue,
    contextValue: context,
    variableValues: variables
  });

  expect(result.data?.ProductAddMutation.productEdge).toBeTruthy();
  expect(result.data?.ProductAddMutation.productEdge.node.name).toEqual(
    'awesome-product'
  );
  expect(result.data?.ProductAddMutation.error).toBeNull();
  expect(
    sanitizeTestObject(result.data, [...defaultFrozenKeys])
  ).toMatchSnapshot();
});
