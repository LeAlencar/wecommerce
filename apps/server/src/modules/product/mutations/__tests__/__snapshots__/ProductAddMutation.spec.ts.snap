// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create an product 1`] = `
{
  "ProductAddMutation": {
    "error": "EMPTY",
    "productEdge": {
      "cursor": "FROZEN-CURSOR",
      "node": {
        "id": "FROZEN-ID",
        "name": "awesome-product",
      },
    },
  },
}
`;
